pkg/netlight/resource/resource.go

```diff
@@ -0,0 +1,399 @@
+		IPNet: ip,
+	}
+	err = netlink.AddrAdd(link, &addr)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("failed to add ip address to link: %w", err)
+	}
+
+	return netlink.LinkSetUp(link)
+}
+
+// Get return resource handler
+func Get(name string) (*Resource, error) {
+	nsName := fmt.Sprintf("n%s", name)
+
+	if namespace.Exists(nsName) {
+		return &Resource{name}, nil
+	}
+
+	return nil, fmt.Errorf("resource not found: %s", name)
+}
+
+var networkResourceNet = net.IPNet{
+	IP:   net.ParseIP("100.64.0.0"),
+	Mask: net.IPv4Mask(0xff, 0xff, 0, 0),
+}
+
+func (r *Resource) AttachPrivate(id string, vmIp net.IP) (device pkg.TapDevice, err error) {
+	nsName := fmt.Sprintf("n%s", r.name)
+	netNs, err := namespace.GetByName(nsName)
+	if err != nil {
+		return
+	}
+	if vmIp[3] == 1 {
+		return device, fmt.Errorf("ip %s is reserved", vmIp.String())
+	}
+
+	var addrs []netlink.Addr
+
+	if err = netNs.Do(func(_ ns.NetNS) error {
+		link, err := netlink.LinkByName(infPrivate)
+		if err != nil {
+			return err
+		}
+		addrs, err = netlink.AddrList(link, netlink.FAMILY_V4)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return
+	}
+	if len(addrs) != 1 {
+		return device, fmt.Errorf("expect addresses on private interface to be 1 got %d", len(addrs))
+	}
+	gw := addrs[0].IPNet
+	if !gw.Contains(vmIp) {
+		return device, fmt.Errorf("ip not in range")
+	}
+
+	deviceName := ifaceutil.DeviceNameFromInputBytes([]byte(id))
+	tapName := fmt.Sprintf("b-%s", deviceName)
+
+	privateNetBr := fmt.Sprintf("r%s", r.name)
+	hw := ifaceutil.HardwareAddrFromInputBytes([]byte(tapName))
+
+	ip := &net.IPNet{
+		IP:   vmIp,
+		Mask: gw.Mask,
+	}
+	_, getLinkErr := netlink.LinkByName(tapName)
+	if getLinkErr != nil {
+		mtap, err := macvtap.CreateMACvTap(tapName, privateNetBr, hw)
+		if err != nil {
+			return pkg.TapDevice{}, err
+		}
+
+		if err = netlink.AddrAdd(mtap, &netlink.Addr{
+			IPNet: ip,
+		}); err != nil {
+			return pkg.TapDevice{}, err
+		}
+	}
+
+	routes := []pkg.Route{
+		{
+			Gateway: gw.IP,
+		},
+		{
+			Net:     networkResourceNet,
+			Gateway: gw.IP,
+		},
+	}
+	return pkg.TapDevice{
+		Name:   tapName,
+		Mac:    hw,
+		IP:     ip,
+		Routes: routes,
+	}, nil
+}
+
+func (r *Resource) AttachMycelium(id string, seed []byte) (device pkg.TapDevice, err error) {
+	nsName := fmt.Sprintf("n%s", r.name)
+	netNS, err := namespace.GetByName(nsName)
+	if err != nil {
+		return
+	}
+	name := filepath.Base(netNS.Path())
+	netSeed, err := os.ReadFile(filepath.Join(myceliumSeedDir, name))
+	if err != nil {
+		return
+	}
+
+	inspect, err := inspectMycelium(netSeed)
+	if err != nil {
+		return
+	}
+
+	ip, gw, err := inspect.IPFor(seed)
+	if err != nil {
+		return
+	}
+
+	deviceName := ifaceutil.DeviceNameFromInputBytes([]byte(id))
+	tapName := fmt.Sprintf("m-%s", deviceName)
+
+	myBr := fmt.Sprintf("m%s", r.name)
+	hw := ifaceutil.HardwareAddrFromInputBytes([]byte(tapName))
+
+	_, err = macvtap.CreateMACvTap(tapName, myBr, hw)
```

Comment:
Same for mycelium this can perfectly be a tap device (not macvtap) hence it becomes simpler to pass to the `cloud-hypervisor` command line

Reply:
macvtaps or macvlans are a disaster if they're more than one with a bridge interface as parent.
Somewhere since that dummy necessity version, macvlans have problems ... what they specifically are, I'm not sure yet, but I suspect seen that macvlans are also bridges in a lighter version, they don't cooperate well with regular bridges

Reply:
I felt like since the https://bugzilla.kernel.org/show_bug.cgi?id=214631 bug, macvlans went worse and worse over time.
probably because of their infrequent use.
let's see how much work it is, but I guess